### DISCUSSION TITLE
Parallel tests with IClassFixture<> from BaseClass

### DISCUSSION THREAD

**Discussion by lucasteles** [xunit/xunit#2571] [Question] [answered]
Hi, 

I need some help, I am trying to reuse some local and class fixtures which will repeat a bunch o times in my integration tests.
- For local fixtures, I inherited from a base test class (Is there a better way? I would love to have an IFixture<> for each test)
- For class fixtures, I used  IClassFixture<>

My class fixture is responsible for setting up and running the database test container and the local fixture is responsible for clearing the test data before each test.

So I had a DbFixture like: 
```csharp
public class DbFixture : IAsyncLifetime
{
    readonly PostgreSqlTestcontainer postgresSql;
    public DbFixture()  {  this.postgresSql = new TestcontainersBuilder<PostgreSqlTestcontainer>(); }
    async Task IAsyncLifetime.InitializeAsync() { /* start container  */ }
    async Task IAsyncLifetime.DisposeAsync() => await postgresSql.DisposeAsync();
    public async Task ClearDbData() { /* set database to a well known state */ }
}

public abstract class AppFixture : IAsyncLifetime,  IClassFixture<DbFixture>
{
    readonly DbFixture dbFixture;
    protected AppFixture(ServicesFixture servicesFixture, DbFixture dbFixture) =>
        this.dbFixture = dbFixture;
   
    Task IAsyncLifetime.InitializeAsync() =>
        return dbFixture.ClearDbData();

    async Task IAsyncLifetime.DisposeAsync()
        await testScope.DisposeAsync();
        await dbScope.DisposeAsync();
    }
}
```
>  _There are more setups and fixtures on the one I am implementing, but I will omit them for brevity_

I am parting with the `IClassFixture<>` documentation premise:
> When to use: when you want to create a single test context and share it among all the tests in the class and have it cleaned up after all the tests in the class have finished.

When I inherited  `AppFixture` from just one class everything worked as expected,  
But when I have two or more test classes that inherit from `AppFixture` the tests crash, it is because they are sharing the same ClassFixture instance which is not what I expected because of _a single test context and shares it among all the tests in the class_  

So, looks like XUnit is considering the base class as a collection not the concrete class with the tests...

Is there something which I can do to it work as I expected? or another way to have this kinda of setup?

**Comment by bradwilson** [comment, marked as answer]
`IClassFixture<T>` shares a single instance of `T` among all the unit tests in the same class. If there are two or more classes which have `IClassFixture<T>`, then _each class will get its own unique instance of `T` that is shared amongst all it test methods_.

My guess is the issue you're hitting is that your `DbFixture` does not support multiple parallel instances of itself. If you want to share a single instance of `DbFixture` among test methods in many classes, then what you want is a [collection fixture](https://xunit.net/docs/shared-context#collection-fixture), not a class fixture.

**Comment by fairking** [comment]
You can setup your `DbFixture` so it initializes a new fresh database with unique name every time it is created (u can add some random number as a postfix to the database name). In this case you can keep using `IClassFixture`. 

Otherwise if you think only one database should exist for all you test cases then you can use [AssemblyFixture](https://stackoverflow.com/a/53143426/1143349). 

You can also use [collection fixture](https://xunit.net/docs/shared-context#collection-fixture) as @bradwilson suggested but if you have multiple collections then you have to have multiple unique databases anyway.
